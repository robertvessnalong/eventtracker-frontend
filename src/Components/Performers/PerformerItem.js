import { Card, Button, Col } from 'react-bootstrap';
import { Link } from 'react-router-dom';

const PerformerItem = ({ performer }) => {
  return (
    <Card className='mt-4 mb-4 p-0 text-white'>
      <Card.Img src={performer.images.huge} alt='Card image' />
      <Card.ImgOverlay className='center-content'>
        <Card.Title style={{ fontWeight: 'bolder' }} as='h2'>
          {performer.name}
        </Card.Title>
        <Col xs={12} md={6}>
          <Link to={`/performers/${performer.id}`} rel='noreferrer'>
            <Button className='blue-btn' variant='primary'>
              See Events
            </Button>
          </Link>
        </Col>
      </Card.ImgOverlay>
    </Card>
  );
};

export default PerformerItem;
