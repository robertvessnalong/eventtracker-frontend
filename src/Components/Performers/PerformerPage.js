import { useState, useEffect, useContext } from 'react';
import { Container, Row, Col, Spinner, Alert } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import UserContext from '../../Context/UserContext';
import EventFinderApi from '../../API/api';
import EventItem from '../Events/EventItem';
import { AiFillHeart, AiOutlineHeart } from 'react-icons/ai';
import uuid from 'react-uuid';

const PerformerPage = () => {
  const { user, setUser, loggedIn } = useContext(UserContext);
  const [performer, setPerformer] = useState([]);
  const [performerEvents, setPerformerEvents] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState([]);
  const [hasEvents, setHasEvents] = useState(false);

  useEffect(() => {
    const getPerformer = async () => {
      const performerId = window.location.href.split('/').pop();
      const performer = await EventFinderApi.getPerformer({ performerId });
      const events = await EventFinderApi.getEvents({ search: performer.name });
      if (performer.error) {
        setError(true);
        setErrorMsg(performer.error);
      } else {
        if (events.length > 0) {
          setHasEvents(false);
          setPerformerEvents(events);
        } else {
          setHasEvents(true);
        }
      }
      setPerformer(performer);
      setLoading(false);
    };
    getPerformer();
  }, []);

  if (loading) {
    return (
      <Row>
        <Col style={{ textAlign: 'center' }}>
          <Spinner animation='border' role='status'>
            <span className='visually-hidden'>Loading...</span>
          </Spinner>
        </Col>
      </Row>
    );
  }

  // Handle Favorties
  const handleFavorites = async (e, performer) => {
    const eventData = {
      type: 'performer',
      performerId: performer.id,
      performerName: performer.name,
      performerImage: performer.image,
      performerUrl: performer.url,
      uuid: user['uuid'],
    };
    const res = e.target.classList.contains('favorited')
      ? await EventFinderApi.removefavorite(eventData)
      : await EventFinderApi.favoriteEvent(eventData);
    if (res.error) {
      console.error(res.error);
    } else {
      const user = JSON.parse(localStorage.getItem('user'));
      const userData = await EventFinderApi.getUserInfo(user.user, user.token);
      setUser(userData);
    }
  };

  return (
    <section className='PerformerPage'>
      <header>
        <h2>{error ? errorMsg : performer.name}</h2>
        {!error && (
          <>
            {loggedIn && (
              <>
                {EventFinderApi.favoriteCheck(user.favorites, performer) ? (
                  <div
                    className='favorite-btn favorited'
                    onClick={(e) => handleFavorites(e, performer)}
                  >
                    <AiFillHeart />
                  </div>
                ) : (
                  <div
                    className='favorite-btn'
                    onClick={(e) => handleFavorites(e, performer)}
                  >
                    <AiOutlineHeart />
                  </div>
                )}
              </>
            )}
          </>
        )}
        <hr></hr>
      </header>
      <Container className='Performer-Events'>
        {hasEvents && (
          <Alert className='mt-4' variant='warning'>
            Sorry, no events found! Try Again
          </Alert>
        )}
        <Row xs={1} xl={2}>
          {performerEvents.map((event) => (
            <Col key={uuid()}>
              <EventItem event={event} favorites={user.favorites} />
            </Col>
          ))}
        </Row>
      </Container>
      <footer>
        <Link className='blue-btn' to='/performers'>
          Go Back
        </Link>
      </footer>
    </section>
  );
};

export default PerformerPage;
