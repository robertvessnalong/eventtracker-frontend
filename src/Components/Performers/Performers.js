import { useState, useEffect, useContext } from 'react';
import {
  Container,
  Row,
  Col,
  Spinner,
  FloatingLabel,
  Form,
  Button,
  Alert,
} from 'react-bootstrap';
import UserContext from '../../Context/UserContext';
import PerformerItem from './PerformerItem';
import EventsFinderApi from '../../API/api';
import uuid from 'react-uuid';

const Performers = () => {
  const formInitalState = {
    search: '',
  };
  const { user } = useContext(UserContext);
  const [page, setPage] = useState(1);
  const [formData, setFormData] = useState(formInitalState);
  const [performers, setPerformers] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState([]);
  const [hasPerformers, setHasPerformers] = useState(false);

  useEffect(() => {
    const getPerformers = async () => {
      const res = await EventsFinderApi.getPerformers({ page });
      setPerformers(res);
      setLoading(false);
    };
    getPerformers();
  }, [page]);

  if (loading) {
    return (
      <Row>
        <Col style={{ textAlign: 'center' }}>
          <Spinner animation='border' role='status'>
            <span className='visually-hidden'>Loading...</span>
          </Spinner>
        </Col>
      </Row>
    );
  }

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await EventsFinderApi.getPerformers(formData);
    if (res.error) {
      setError(true);
      setErrorMsg(res.error);
    } else {
      if (performers.length > 0) {
        setHasPerformers(false);
        setFormData(formInitalState);
        setPerformers(res);
      } else {
        setHasPerformers(true);
      }
    }
  };

  const handlePagination = (e) => {
    if (e.target.classList.contains('btn')) {
      const attr = e.target.getAttribute('aria-label');
      if (attr === 'Previous' && page !== 1) {
        setPage((prevState) => prevState - 1);
      } else if (attr === 'Next' && page >= 1) {
        setPage((prevState) => prevState + 1);
      }
    }
  };

  return (
    <div className='Performers'>
      <header>
        <h2>Performers</h2>
        <span>See Your Favorite Performers Show</span>
        <hr></hr>
      </header>
      <Container>
        {hasPerformers && (
          <Alert className='mt-4' variant='warning'>
            Sorry, no performers found! Try Again
          </Alert>
        )}
        <div className='Search'>
          <Form onSubmit={handleSubmit} className='mt-4'>
            {error && <Alert variant='danger'>{errorMsg}</Alert>}
            <Row className='g-2'>
              <Col md>
                <FloatingLabel
                  controlId='floatingInputGrid'
                  label='Search Performers...'
                >
                  <Form.Control
                    type='text'
                    name='search'
                    value={formData.search}
                    onChange={handleChange}
                  />
                </FloatingLabel>
              </Col>
              <Col md={3}>
                <Button
                  type='submit'
                  className='h-100 w-100 blue-btn'
                  variant='primary'
                >
                  Find Performer
                </Button>{' '}
              </Col>
            </Row>
          </Form>
        </div>
      </Container>
      <Row xs={1} lg={2} xl={3}>
        {performers.map((performer) => (
          <Col key={uuid()}>
            <PerformerItem performer={performer} favorites={user.favorites} />
          </Col>
        ))}
      </Row>
      <Row className='mt-5'>
        <Col>
          <div aria-label='pagination'>
            <ul
              className='pagination justify-content-center'
              key={uuid()}
              onClick={handlePagination}
            >
              <Button
                disabled={page === 1 ? true : false}
                className='page-link'
                href='#'
                aria-label='Previous'
                key={uuid()}
              >
                <span className='page-arrow' aria-hidden='true'>
                  &laquo;
                </span>
              </Button>
              <Button
                className='page-link'
                href='#'
                aria-label='Next'
                key={uuid()}
              >
                <span className='page-arrow' aria-hidden='true'>
                  &raquo;
                </span>
              </Button>
            </ul>
          </div>
        </Col>
      </Row>
    </div>
  );
};

export default Performers;
