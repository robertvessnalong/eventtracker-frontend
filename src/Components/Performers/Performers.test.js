import {
  render,
  screen,
  fireEvent,
  waitFor,
  waitForElementToBeRemoved,
  getByPlaceholderText,
  getByLabelText,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Performers from './Performers';
import TestUser from '../../Testing/TestUser';
import UserContext from '../../Context/UserContext';
import { BrowserRouter } from 'react-router-dom';

describe('Events Component', () => {
  // Smoke Test
  test('renders without crashing', async () => {
    render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Performers />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitFor(() => {
      const headerSpan = screen.getByText(/See Your Favorite Performers Show/i);
      expect(headerSpan).toBeInTheDocument();
    });
  });
  // Smoke Test
  test('Alert Not Rendered', async () => {
    render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Performers />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitFor(() => {
      const alertMsg = screen.queryByText(/Sorry, no events found! Try Again/i);
      expect(alertMsg).not.toBeInTheDocument();
    });
  });

  test('fireSubmit', async () => {
    const { getByText, getByRole } = render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Performers />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitForElementToBeRemoved(screen.getByText('Loading...'));
    await waitFor(() => {
      fireEvent.change(getByRole('textbox', { name: /search/i }), {
        query: '',
      });
      fireEvent.submit(getByText('Find Performer'));
    });
    const alertMsg = screen.getByText(/Billy Joel/i);
    expect(alertMsg).toBeInTheDocument();
  });

  // Snapshot
  test('matches snapshot', async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <Performers />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitForElementToBeRemoved(screen.getByText('Loading...'));
    expect(asFragment()).toMatchSnapshot();
  });
});
