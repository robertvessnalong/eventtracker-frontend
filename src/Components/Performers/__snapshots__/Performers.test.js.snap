// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Performers"
  >
    <header>
      <h2>
        Performers
      </h2>
      <span>
        See Your Favorite Performers Show
      </span>
      <hr />
    </header>
    <div
      class="container"
    >
      <div
        class="Search"
      >
        <form
          class="mt-4"
        >
          <div
            class="g-2 row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-floating"
              >
                <input
                  class="form-control"
                  id="floatingInputGrid"
                  name="search"
                  type="text"
                  value=""
                />
                <label
                  for="floatingInputGrid"
                >
                  Search Performers...
                </label>
              </div>
            </div>
            <div
              class="col-md-3"
            >
              <button
                class="h-100 w-100 blue-btn btn btn-primary"
                type="submit"
              >
                Find Performer
              </button>
               
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="row row-cols-xl-3 row-cols-lg-2 row-cols-1"
    >
      <div
        class="col"
      >
        <div
          class="mt-4 mb-4 p-0 text-white card"
        >
          <img
            alt="Card image"
            class="card-img"
            src="https://seatgeek.com/images/performers-landscape/billy-joel-82da62/303/huge.jpg"
          />
          <div
            class="center-content card-img-overlay"
          >
            <h2
              class="card-title"
              style="font-weight: bolder;"
            >
              Billy Joel
            </h2>
            <div
              class="col-md-6 col-12"
            >
              <a
                href="/performers/303"
                rel="noreferrer"
              >
                <button
                  class="blue-btn btn btn-primary"
                  type="button"
                >
                  See Events
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mt-5 row"
    >
      <div
        class="col"
      >
        <div
          aria-label="pagination"
        >
          <ul
            class="pagination justify-content-center"
          >
            <a
              aria-disabled="true"
              aria-label="Previous"
              class="page-link btn btn-primary disabled"
              role="button"
            >
              <span
                aria-hidden="true"
                class="page-arrow"
              >
                «
              </span>
            </a>
            <a
              aria-label="Next"
              class="page-link btn btn-primary"
              href="#"
              role="button"
              tabindex="0"
            >
              <span
                aria-hidden="true"
                class="page-arrow"
              >
                »
              </span>
            </a>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
